### ISSUE TITLE
CSS.highlights.highlightsFromPoint API

### ISSUE THREAD

**Issue by ffiori** [w3ctag/design-reviews#1043]
こんにちは TAG-さん!
 
I'm requesting a TAG review of highlightsFromPoint API.

The highlightsFromPoint() API is designed to enhance user interaction with highlighted content on web pages. Highlights are used to visually distinguish specific ranges within a document, such as comments, search results, or spelling errors. This API aims to provide a robust mechanism for identifying and interacting with these highlights, addressing challenges related to multiple overlapping highlights, shadow DOM handling, and performance optimization. More details were discussed in the original issue filed in csswg: https://github.com/w3c/csswg-drafts/issues/7513.

  - Explainer¹: https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/highlight/HighlightsFromPointsExplainer.md
  - Specification: https://github.com/w3c/csswg-drafts/pull/11507
  - WPT Tests: In progress https://chromium-review.googlesource.com/c/chromium/src/+/6074351
  - User research: N/A
  - Security and Privacy self-review²: N/A as there are no additional security implications over existing API.
  - GitHub repo: https://github.com/w3c/csswg-drafts/issues
  - Primary contacts:
      - Fernando Fiori ([@ffiori](https://github.com/ffiori)), Microsoft, implementer
  - Organization/project driving the design: Microsoft
  - Multi-stakeholder feedback³: 
      - Chromium comments: https://chromestatus.com/feature/4552801607483392
      - Mozilla comments: https://github.com/mozilla/standards-positions/issues/1068
      - WebKit comments: https://github.com/WebKit/standards-positions/issues/394
  - Status/issue trackers for implementations⁴: https://chromestatus.com/feature/4552801607483392

Further details:

  - [x] I have reviewed the TAG's [Web Platform Design Principles](https://www.w3.org/TR/design-principles/)
  - Previous early design review, if any: N/A
  - Relevant time constraints or deadlines: None
  - The group where the incubation/design work on this is being done (or is intended to be done in the future): CSSWG
  - The group where standardization of this work is intended to be done ("unknown" if not known): CSSWG
  - Major unresolved issues with or opposition to this design: None
  - This work is being funded by: Microsoft

**Comment by xiaochengh** [comment]
Hi @ffiori - thanks for sending this our way. We discussed in a TAG breakout meeting this morning. One question we had was regarding the choice to make this a synchronous API vs using something like an event model. We are concerned about the performance impact of using a synchronous approach. Have you considered asynchronous alternatives like firing an event when user interacts with a custom highlight?

**Comment by ffiori** [comment]
Hi @xiaochengh, thanks for discussing the topic.

The reason for implementing `highlightsFromPoint` has been discussed with the CSSWG [here](https://github.com/w3c/csswg-drafts/issues/7513#issuecomment-1248320696) before. Just wanted to make clear that by implementing this API we're not discarding implementing an event-based alternative in the future if problems arise or we find other use cases that could apply to that specific scenario.

Summarizing, we chose to move forward with `highlightsFromPoint` because it satisfies various use cases (for example, interactions with custom spellcheck, annotation popups over words, interactions with custom find-on-page, to name a few), and it's a much simpler approach than defining a whole new model for events (as was discussed in [the same issue linked above](https://github.com/w3c/csswg-drafts/issues/7513)).

Let me know if you'd like me to expand on some parts of my comment or if there's any other questions.

**Comment by xiaochengh** [comment]
@ffiori Thanks for the prompt reply!

I see this hit-testing API has been compared with an event API, and major reason the CSSWG moved away from it due to its much higher complexity. Is that correct, and did I miss any other points?

Btw it will be great if the event API is listed in the [Alternative Solutions](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/highlight/HighlightsFromPointsExplainer.md#alternative-solutions) section of the explainer with a comparison.

**Comment by ffiori** [comment]
Hi @xiaochengh, thanks for the quick reply. You're right about the reason for choosing this API over the event API.

I just updated the explainer as suggested. Let me know if you have any other questions or concerns.

**Comment by xiaochengh** [comment]
Hi @ffiori, the TAG discussed it in a breakout today and are still concerned with adding an API that may block the main thread and cause layout thrashing. Per [Design Principles](https://www.w3.org/TR/design-principles/#synchronous), synchronous APIs should have short execution time. Are there other asynchronous alternatives that have been considered?

**Comment by ffiori** [comment]
Hi @xiaochengh, thanks for following up on this topic by filing an issue. One of the reasons why we found this API to be worth pursuing is because it maintains the current patterns for developers (like other *FromPoint APIs, i.e. `elementFromPoint`, `caretPositionFromPoint`) so it's easier and intuitive for them to adopt it.

It's also up to the developers to call the API with a clean layout. We can give them some guidance by adding a note in the spec and in MDN mentioning this. We can also give an example like the following:

```
document.addEventListener('click', (event) => {
    requestAnimationFrame(() => {
        requestAnimationFrame(() => {
            const highlights = CSS.highlights.highlightsFromPoint(event.clientX, event.clientY);
            // Do something with the highlights
            console.log(highlights);
        });
    });
});
```

**Comment by xiaochengh** [comment]
Hi @ffiori, the TAG discussed it and still found our concerns unresolved. We understand that this API is the simplest solution to the target use case, but we are concerned with synchronous layout-performing APIs in general, and the introduction of another API in this class might not be the good way to go.

1. Synchronous layout-performing APIs, on their own, violates design principle [be synchronous when appropriate](https://www.w3.org/TR/design-principles/#synchronous):

> An API should generally be synchronous if the following rules of thumb apply:
> - ...
> - The execution time is short and deterministic.

2. There's no way to reliably prevent layout thrashing caused by such APIs, as otherwise we wouldn't have invented Intersection Observer.

3. Consistency with existing layout-performing APIs does not justify the introduction of another one. We should [leave the web better than you found it](https://www.w3.org/TR/design-principles/#leave-the-web-better), and in particular:

> Issues that are present with a certain web technology now may be fixed in a subsequent iteration. Duplicating these issues makes fixing them more difficult. By adhering to this principle we can make sure overall platform quality improves over time.

For now we are closing this review. We still encourage the exploration of asynchronous alternatives, be it events, observers, Promises, ...

**Comment by zcorpan** [comment]
I commented in https://github.com/mozilla/standards-positions/issues/1068#issuecomment-2788724258

> I think there is value in consistency with the existing `*FromPoint` APIs (which also flush layout).
> 
> If we want to provide a mechanism to use APIs at appropriate times without flushing layout, I think that should be done holistically for all layout-flushing APIs.

**Comment by ffiori** [comment]
Thank you again for the review, I really appreciate the feedback. I understand the concern, but I agree with @zcorpan that the sync layout issue could be addressed separately and keep this API as developer-friendly as possible.

One possible direction could be Alex Russell’s proposal on event phases, which aims to address this more holistically: https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/EventPhases/explainer.md.

Also, in the Alternative Solutions section of [the highlightsFromPoint explainer](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/highlight/HighlightsFromPointsExplainer.md#alternative-solutions), I’ve expanded on other async options and explained why they’re less attractive than the current proposal.

**Comment by xiaochengh** [comment]
Thanks for the feedback!

**Comment by xiaochengh** [comment]
Hi @ffiori, the TAG discussed it at a Breakout today. While we still have the same concern of introducing another potentially layout-thrashing API, we agree that solving the issue of layout-thrashing is out of the scope of this particular issue, and a more holistic solution is needed.

Thank you for referring to the [event phases proposal](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/EventPhases/explainer.md). We are happy to see, and would like to encourage more explorations into the issue of layout thrashing as a whole.
